exonum/src/events/noise/mod.rs

```diff
@@ -40,80 +41,109 @@ pub struct HandshakeParams {
-                .and_then(|(len, buf)| write(stream, &buf, len))
-                .and_then(move |(stream, _msg)| {
-                    let noise = noise.into_transport_mode()?;
-                    let framed = stream.framed(MessagesCodec::new(max_message_len, noise));
-                    Ok(framed)
-                })
-        });
-
-    Box::new(framed)
+impl Handshake for NoiseHandshake {
+    fn listen<S>(self, stream: S) -> HandshakeResult<S>
+    where
+        S: AsyncRead + AsyncWrite + 'static,
+    {
+        let framed = self.read_handshake_msg(stream)
+            .and_then(|(stream, handshake)| handshake.write_handshake_msg(stream))
+            .and_then(|(stream, handshake)| handshake.read_handshake_msg(stream))
+            .and_then(|(stream, handshake)| handshake.finalize(stream));
+        Box::new(framed)
+    }
+
+    fn send<S>(self, stream: S) -> HandshakeResult<S>
+    where
+        S: AsyncRead + AsyncWrite + 'static,
+    {
+        let framed = self.write_handshake_msg(stream)
+            .and_then(|(stream, handshake)| handshake.read_handshake_msg(stream))
+            .and_then(|(stream, handshake)| handshake.write_handshake_msg(stream))
+            .and_then(|(stream, handshake)| handshake.finalize(stream));
+        Box::new(framed)
+    }
 }
 
-fn read(sock: TcpStream) -> Box<Future<Item = (TcpStream, Vec<u8>), Error = io::Error>> {
+fn read<S: AsyncRead + 'static>(sock: S) -> impl Future<Item = (S, Vec<u8>), Error = io::Error> {
     let buf = vec![0u8; HANDSHAKE_HEADER_LENGTH];
-    Box::new(
-        read_exact(sock, buf)
-            .and_then(|(stream, msg)| read_exact(stream, vec![0u8; msg[0] as usize])),
-    )
+    // First byte of handshake message is payload length, remaining bytes [1; len] is
+    // the handshake payload. Therefore, we need to read first byte and after that
+    // remaining payload.
+    read_exact(sock, buf).and_then(|(stream, msg)| read_exact(stream, vec![0u8; msg[0] as usize]))
 }
 
-fn write(
-    sock: TcpStream,
+fn write<S: AsyncWrite + 'static>(
+    sock: S,
     buf: &[u8],
     len: usize,
-) -> Box<Future<Item = (TcpStream, Vec<u8>), Error = io::Error>> {
-    let mut message = vec![0u8; HANDSHAKE_HEADER_LENGTH];
-    LittleEndian::write_u16(&mut message, len as u16);
-    message.extend_from_slice(&buf[0..len]);
-    Box::new(write_all(sock, message))
-}
+) -> impl Future<Item = (S, Vec<u8>), Error = io::Error> {
+    if len > NOISE_MAX_HANDSHAKE_MESSAGE_LENGTH {
+        return Either::A(err(io::Error::new(
+            io::ErrorKind::Other,
+            "Message size exceeds max handshake message size",
+        )));
+    }
 
-fn write_handshake_msg(
-    noise: &mut NoiseWrapper,
-) -> Box<Future<Item = (usize, Vec<u8>), Error = io::Error>> {
-    let res = noise.write_handshake_msg();
-    Box::new(done(res.map_err(|e| e.into())))
+    let mut message = vec![len as u8; HANDSHAKE_HEADER_LENGTH];
+    message.extend_from_slice(&buf[0..len]);
```

Comment:
I wonder if this would work without reallocation. Are two invocations of `write_all` equivalent to a single invocation on the buffer concatenation?

Reply:
It's possible to use two `write_all` invocations.